library/gcm.c

```diff
@@ -49,6 +49,52 @@ void mbedtls_gcm_init(mbedtls_gcm_context *ctx)
     memset(ctx, 0, sizeof(mbedtls_gcm_context));
 }
 
+#if defined(MBEDTLS_GCM_LARGETABLE)
+
+static inline void mbedtls_xor16_rev(void *r, const void *a, const void *b)
+{
+  uint32_t *r32 = (uint32_t *) r;
```

Comment:
Can this be replaced by using `mbedtls_xor` instead (or maybe `mbedtls_xor_no_simd` - depending on how it's being used, this may be faster)?

Reply:
In a quick local test on M1 Pro, `mbedtls_xor` is a lot slower, but `mbedtls_xor_no_simd` is a further 10% improvement.

Reply:
This is in optimization mainly aimed at emedded systems that allow 32bit unaligned (on byte level) memory access. For various cryptographic implementations. Here, XORs are used during every iteration when processing data. The mbedtls_xor function is designed for arbitrary length, hence, additional instructions are executed in every processing loop to check the length. As our block ciphers are fixed to 16 bytes, we already know how much data has to be xored during every iteration. Hence, we can skip the length check instructions. It is also an optimization that helps in other implementations such as AES CTR. Maybe, we can move the mbedtls_xor16 function in to a more generic header flie. As it is very short, inlineing also helps. Maybe some compilers are intelligent enough to realize that always 16 bytes have to be processed to optimize the inlined code, however, the compilers we use for our embedded systems do not. In addition, I assume in the implementation that unaligned memory access is possible, if that is not the case, we would have to fallback to an implementation that works with single bytes as it is also done in the mbedtls_xor function.

Using mbedtls_xor everywhere:
```
  AES-GCM-128              :     446380 KiB/s,          4 cycles/byte
  AES-GCM-192              :     424312 KiB/s,          4 cycles/byte
  AES-GCM-256              :     415949 KiB/s,          4 cycles/byte
```
For 64-bit Intel architecture, that is better than my mbedtls_xor16 function. Most likely as I am accessing memory on a 4-byte basis and mbedtls_xor is taking 8-byte at a time.

Replacing mbedtls_xor16 by mbedtls_xor, still using mbedtls_xor16_rev:
```
  AES-GCM-128              :     440217 KiB/s,          4 cycles/byte
  AES-GCM-192              :     422093 KiB/s,          5 cycles/byte
  AES-GCM-256              :     414839 KiB/s,          4 cycles/byte
```
Has not much effect on performance, as mbedtls_xor16_rev is only used for during context initialization.

Replacing mbedtls_xor16_rev by mbedtls_xor, still using mbedtls_xor16:
```
  AES-GCM-128              :     395110 KiB/s,          5 cycles/byte
  AES-GCM-192              :     385166 KiB/s,          5 cycles/byte
  AES-GCM-256              :     371789 KiB/s,          5 cycles/byte
```
mbedtls_xor16 is used in every loop, as my implementation is slower than the mbedtls_xor variant, the performance decreases.

Replacing mbedtls_xor16_rev by mbedtls_xor, replacing mbedtls_xor16 by mbedtls_xor16_64:
```
  AES-GCM-128              :     447553 KiB/s,          4 cycles/byte
  AES-GCM-192              :     433143 KiB/s,          4 cycles/byte
  AES-GCM-256              :     418608 KiB/s,          5 cycles/byte
```
This results in a slightly better performance than using mbedtls_xor.

mbedtls_xor16_64 is defined as follows:
```
static inline void mbedtls_xor16_64(void *r, const void *a, const void *b)
{
    uint64_t *r64 = (uint64_t *) r;
    const uint64_t *a64 = (const uint64_t *) a;
    const uint64_t *b64 = (const uint64_t *) b;
    r64[0] = a64[0] ^ b64[0];
    r64[1] = a64[1] ^ b64[1];
}
```

Reply:
There is another reason why I require the _rev variant of that function:
```
    for (i = 2; i < 256; i <<= 1) {
        for (j = 1; j < i; j++) {
            mbedtls_xor16_rev(H[i+j], H[i], H[j]);
        }
    }
```
As source and target arrays are the same, it only works in this situation when I start XORing from back to front.

Reply:
Under gcc and clang, calling `mbedtls_xor` or `mbedtls_xor_no_simd` with a constant length argument ends up optimising away all the conditionals & length checks, resulting in very performant code; similarly for alignment, it will do the best thing depending on whether unaligned access is possible or not. I did look at a `mbedtls_xor16` variant when implementing these, and concluded (for gcc and clang) that there was no benefit at all.

It seems that IAR doesn't do a very good job here - are you using IAR? If that's the case, I'd prefer to fix these functions for IAR (and get the perf win everywhere) rather than special-case these functions in `gcm`.

Reply:
I've prototyped an IAR fix here - please take a look https://github.com/Mbed-TLS/mbedtls/pull/8719

Reply:
We are currently stuck on some older GCC releases due to OS dependencies and so on. However, I will run some more tests to find a good compromize and then I will update this PR.

Reply:
I'd be interested to know the gcc version?

Reply:
I seem to recall you were using gcc 4.5.2? Given the age of this compiler, I don't think we would accept performance optimisations that only benefit very old versions.

Reply:
I tried #8719 with "GCC (eCosCentric GNU tools 7.3.0d) 7.3.0" and improves the performance as expected.

Reply:
I have to correct myself. After a clean built, I realized that the patches in #8719 bring only minimal improvements. However, if I replace memcpy operations by dereferencing pointers similar to IAR, I get a significant performance increase of over 100%.

Reply:
I can reproduce using `gcc-9`. It looks like the problem is that `mbedtls_xor_no_simd` does not get inlined. When I add:
```
#if defined(__IAR_SYSTEMS_ICC__)
#pragma inline = forced
#elif defined(__GNUC__)
__attribute__((always_inline))
#endif
```

in front of `mbedtls_xor_no_simd` in `common.h`, the performance regression is resolved. Do you see the same?

Reply:
Actually, all versions of gcc are affected up to at least gcc 12.

Reply:
Why is there even a no_simd version of mbedtls_xor? Should it either always be faster or slower than the regular mbedtls_xor?

Reply:
When you need to access the results immediately afterwards using non-SIMD instructions, it creates latency to transfer the result from the SIMD registers into the scalar registers, and ends up being slower. If that's not needed, the SIMD version is faster. So the preferred option depends on exactly how it's being used.

Reply:
That makes sense. Unfortunately, the compilers due not figure that out on their own and automatically choose the appropriate solution.

Reply:
Using the mbedtls_xor from #8719 instead of mbedtls_xor16 with "GCC (eCosCentric GNU tools 7.3.0d) 7.3.0" on my embedded hardware, reduces the throughput by almost 50 %:

with mbedtls_xor16:
AES-GCM-128              :       1111 KiB/s,         87 cycles/byte
AES-GCM-192              :       1105 KiB/s,         88 cycles/byte
AES-GCM-256              :       1099 KiB/s,         88 cycles/byte

with mbedtls_xor from #8719:
AES-GCM-128              :        599 KiB/s,        163 cycles/byte
AES-GCM-192              :        597 KiB/s,        163 cycles/byte
AES-GCM-256              :        595 KiB/s,        164 cycles/byte

Reply:
It seems to depend on the optimizations:
with mbedtls_xor16, -Os:
AES-GCM-128 : 1111 KiB/s, 87 cycles/byte
AES-GCM-192 : 1105 KiB/s, 88 cycles/byte
AES-GCM-256 : 1099 KiB/s, 88 cycles/byte

with mbedtls_xor from https://github.com/Mbed-TLS/mbedtls/pull/8719, -Os:
AES-GCM-128 : 599 KiB/s, 163 cycles/byte
AES-GCM-192 : 597 KiB/s, 163 cycles/byte
AES-GCM-256 : 595 KiB/s, 164 cycles/byte

with mbedtls_xor16, -O3:
AES-GCM-128              :       1378 KiB/s,         70 cycles/byte
AES-GCM-192              :       1369 KiB/s,         71 cycles/byte
AES-GCM-256              :       1360 KiB/s,         71 cycles/byte

with mbedtls_xor from https://github.com/Mbed-TLS/mbedtls/pull/8719, -O3:
AES-GCM-128              :       1343 KiB/s,         72 cycles/byte
AES-GCM-192              :       1335 KiB/s,         73 cycles/byte
AES-GCM-256              :       1327 KiB/s,         73 cycles/byte

So with active optimizations, the difference is negligible, but still there.

Reply:
I think in this case, you need `mbedtls_xor_no_simd`, as well as the latest update to #8719 which improves unaligned access on old versions of gcc.

Reply:
OK, so that will make no difference in your case, but for users who do have SIMD instructions, they will get better performance from `mbedtls_xor_no_simd` (I've seen this locally testing on M1 Pro).

Reply:
Still it seems like the mbedtls_xor gets only optimized as expected when -O3 is turned on. Even -O2 is not sufficient.

Reply:
For the tests I was still using the code I proposed here: https://github.com/Mbed-TLS/mbedtls/pull/8719#issuecomment-1900097578
However, I am now rebuilding and testing with your latest code.

Reply:
I think we might have an include problem. MBEDTLS_COMPILER_IS_GCC gets defined in common.h, but alignment.h which checks for MBEDTLS_COMPILER_IS_GCC is included into common.h. So the GCC specific code never gets active.

Reply:
if I redefine MBEDTLS_COMPILER_IS_GCC at the beginning of alignment.h, I get an unknown pragma error:
`error: ignoring #pragma language`
from:
```
#if defined(UINT_UNALIGNED)
#pragma language=restore
#endif
```

Reply:
Sorry, thanks for spotting. I've pushed a fix.

Reply:
now replaced by mbedtls_xor_no_simd